File: frontend/src/lib/shared/normalizers/__tests__/name-normalizer.test.ts
import { describe, expect, it } from 'vitest';
import { normalizeClientName } from '../name-normalizer';

describe('normalizeClientName', () => {
  it('removes accents from names', () => {
    expect(normalizeClientName('Café')).toBe('cafe');
    expect(normalizeClientName('niño')).toBe('nino');
    expect(normalizeClientName('François')).toBe('francois');
  });
  
  it('handles complex Unicode characters', () => {
    expect(normalizeClientName('Zürich')).toBe('zurich');
    expect(normalizeClientName('São Paulo')).toBe('saopaulo');
    expect(normalizeClientName('Köln')).toBe('koln');
  });
  
  it('converts to lowercase', () => {
    expect(normalizeClientName('test')).toBe('test');
    expect(normalizeClientName('MiXeD')).toBe('mixed');
  });
  
  it('removes special characters and spaces', () => {
    expect(normalizeClientName('ABC & Co.')).toBe('abcco');
    expect(normalizeClientName('Test-Name')).toBe('testname');
    expect(normalizeClientName('Name (with) [brackets]')).toBe('namewithbrackets');
    expect(normalizeClientName('Name @ Company')).toBe('namecompany');
  });
  
  it('handles null and undefined values', () => {
    expect(normalizeClientName(null as any)).toBeNull();
    expect(normalizeClientName(undefined as any)).toBeNull();
  });
  
  it('handles empty strings', () => {
    expect(normalizeClientName('')).toBeNull();
  });
  
  it('removes all whitespace', () => {
    expect(normalizeClientName('  spaced   out  ')).toBe('spacedout');
    expect(normalizeClientName('line\nbreak')).toBe('linebreak');
    expect(normalizeClientName('tab\tseparated')).toBe('tabseparated');
  });
  
  it('keeps only alphanumeric characters', () => {
    expect(normalizeClientName('Test123')).toBe('test123');
    expect(normalizeClientName('!@#$%^&*()')).toBeNull();
  });
});

describe('normalizeClientName with object input', () => {
  it('normalizes name field and adds normalized_name', () => {
    const input = { name: 'Café René' };
    const result = normalizeClientName(input);
    
    expect(result).toEqual({
      name: 'Café René',
      normalized_name: 'caferene'
    });
  });
  
  it('preserves other fields', () => {
    const input = { 
      name: 'Test Company',
      id: '123',
      type: 'client'
    };
    const result = normalizeClientName(input);
    
    expect(result).toEqual({
      name: 'Test Company',
      normalized_name: 'testcompany',
      id: '123',
      type: 'client'
    });
  });
  
  it('returns input unchanged if no name field', () => {
    const input = { id: '123', type: 'client' };
    const result = normalizeClientName(input);
    
    expect(result).toEqual(input);
  });
  
  it('handles empty name in object', () => {
    const input = { name: '' };
    const result = normalizeClientName(input);
    
    expect(result).toEqual({
      name: '',
      normalized_name: null
    });
  });
});

File: test/lib/shared/normalizers/name_normalizer_test.rb
require "test_helper"

class NameNormalizerTest < ActiveSupport::TestCase
  test "removes accents from names" do
    assert_equal "CAFE", Shared::Normalizers::NameNormalizer.normalize("Café")
    assert_equal "NINO", Shared::Normalizers::NameNormalizer.normalize("niño")
    assert_equal "FRANCOIS", Shared::Normalizers::NameNormalizer.normalize("François")
  end

  test "handles complex Unicode characters" do
    assert_equal "ZURICH", Shared::Normalizers::NameNormalizer.normalize("Zürich")
    assert_equal "SAOPAULO", Shared::Normalizers::NameNormalizer.normalize("São Paulo")
    assert_equal "KOLN", Shared::Normalizers::NameNormalizer.normalize("Köln")
  end

  test "converts to uppercase" do
    assert_equal "TEST", Shared::Normalizers::NameNormalizer.normalize("test")
    assert_equal "MIXED", Shared::Normalizers::NameNormalizer.normalize("MiXeD")
  end

  test "removes special characters and spaces" do
    assert_equal "ABCCO", Shared::Normalizers::NameNormalizer.normalize("ABC & Co.")
    assert_equal "TESTNAME", Shared::Normalizers::NameNormalizer.normalize("Test-Name")
    assert_equal "NAMEWITHBRACKETS", Shared::Normalizers::NameNormalizer.normalize("Name (with) [brackets]")
    assert_equal "NAMECOMPANY", Shared::Normalizers::NameNormalizer.normalize("Name @ Company")
  end

  test "handles nil values" do
    assert_nil Shared::Normalizers::NameNormalizer.normalize(nil)
  end

  test "handles empty strings" do
    assert_nil Shared::Normalizers::NameNormalizer.normalize("")
  end

  test "removes all whitespace" do
    assert_equal "SPACEDOUT", Shared::Normalizers::NameNormalizer.normalize("  spaced   out  ")
    assert_equal "LINEBREAK", Shared::Normalizers::NameNormalizer.normalize("line\nbreak")
    assert_equal "TABSEPARATED", Shared::Normalizers::NameNormalizer.normalize("tab\tseparated")
  end

  test "keeps only alphanumeric characters" do
    assert_equal "TEST123", Shared::Normalizers::NameNormalizer.normalize("Test123")
    assert_nil Shared::Normalizers::NameNormalizer.normalize("!@#$%^&*()")
  end
end
